docs/source/kinds_of_types.rst

```diff
@@ -345,50 +345,176 @@ narrow down the type to a specific type:
+       if x and y:
+           # Both x and y are not None here
+           return x + y
+       else:
+           return None
+
+Sometimes mypy doesn't realize that a value is never ``None``. This notably
+happens when a class instance can exist in a partially defined state,
+where some attribute is initialized to ``None`` during object
+construction, but a method assumes that the attribute is no longer ``None``. Mypy
+will complain about the possible ``None`` value. You can use
+``assert x is not None`` to work around this:
+
+.. code-block:: python
+
+   class Resource:
+       path: Optional[str] = None
+
+       def initialize(self, path: str) -> None:
+           self.path = path
+
+       def read(self) -> str:
+           # We require that the object has been initizlied.
+           assert self.path is not None
+           with open(self.path) as f:  # OK
+              return f.read()
+
+   r = Resource()
+   r.initialize('/foo/bar')
+   r.read()
 
 When initializing a variable as ``None``, ``None`` is usually an
 empty place-holder value, and the actual value has a different type.
-This is why you need to annotate an attribute in a case like this:
+This is why you need to annotate an attribute in a cases like the class
+``Resource`` above:
 
 .. code-block:: python
 
-    class A:
+    class Resource:
+        path: Optional[str] = None
+        ...
+
+This also works for attributes defined within methods:
+
+.. code-block:: python
+
+    class Counter:
         def __init__(self) -> None:
-            self.count = None  # type: int
+            self.count: Optional[int] = None
+
+As a special case, you can use a non-optional type when initializing an
+attribute to ``None`` inside a class body *and* using a type comment,
+since when using a type comment, an initializer is syntacticaly required,
+and ``None`` is used as a dummy, placeholder initializer:
 
-Mypy will complain if you omit the type annotation, as it wouldn't be
-able to infer a non-trivial type for the ``count`` attribute
-otherwise.
+.. code-block:: python
+
+   from typing import List
+
+   class Container:
+       items = None  # type: List[str]  # OK (only with type comment)
 
 Mypy generally uses the first assignment to a variable to
 infer the type of the variable. However, if you assign both a ``None``
-value and a non-``None`` value in the same scope, mypy can often do
-the right thing:
+value and a non-``None`` value in the same scope, mypy can usually do
+the right thing without an annotation:
 
 .. code-block:: python
 
    def f(i: int) -> None:
-       n = None  # Inferred type int because of the assignment below
+       n = None  # Inferred type Optional[int] because of the assignment below
        if i > 0:
             n = i
        ...
 
-Often it's useful to know whether a variable can be
+.. note::
+
+    ``None`` is also used as the return type for functions that don't
+    return a value, i.e. that implicitly return ``None``. (Mypy doesn't
```

Comment:
PEP 484 already specify this, I am not sure we need to discuss this here again (on the other hand there was a recent issue that proposed to improve some wording there).

Reply:
Do you expect mypy users to read through PEP 484 (which is rather long and complex)? Especially for a fairly basic thing like this.

Reply:
I'd like the mypy documentation to be self-contained and PEP 484 should be strictly optional reading.

Reply:
Then I would reformulate this to make message more clear. Currently it starts from mentioning return type, but then  moves to something else. Also I would not mention `NoneType` at all, it can only add to the confusion, I would just re-iterate that for simplicity in `None` in type context means a type with a single value `None`. And separately mention that Python (this is not specific to mypy) returns `None` from a function that either reaches its end without return statement, or has a bare return.